docs/syntax/transforms.md

```diff
@@ -0,0 +1,219 @@
+the field option `init` (see section on interfaces for what that means).
+
+```python
+def execute(self, ctx, tools, iface):
+```
+> The execute method defines the process to produce tranform outputs from
+transform inputs. The three arguments are as follows:
+> 1. `ctx` the Blockwork context object, which has many useful values and
+     methods.
+> 2. `tools` the tools that are mapped into the container
+> 3. `iface` the values provided by the interfaces, mapped into the container.
+
+
+```python
+copy = tools.bash.get_action("cp")
+```
+> Retrieves an action named `cp` from the bash tool
+
+```python
+yield copy(ctx, frm=iface.frm, to=iface.to)
+```
+> Reads the values from the `frm` and `to` interfaces, and passes them through
+to the copy action. Actions must be yielded in order to run.
+
+
+
+## Basic Use
+
+We use transforms by instantiating them, and calling the run method as follows:
+
+```python
+tf = Copy(frm=Path('/my/input/path'), to=Path('my/output/path'))
+# Where ctx is the blockwork context object
+tf.run(ctx)
+```
+
+## Interfaces
+
+Interfaces are defined with a *name*, a *type*, a *direction*, and *options*.
+The *name* and *direction* are trivial and indicated by the property name, and
+the `IN` and `OUT` in `Transform.IN()` and `Transform.OUT()` respectively. The
+*type* and *options* have more depth and will be the focus of this
+section, starting with options.
+
+### Options
+
+Options are passed in through the `IN` and `OUT` transform methods. They will
+are best described by example.
+
+```python
+files: dict[str, Path] = Transform.IN()
+```
+> An input interface which accepts a dictionary of paths and exposes them
+inside the container
+
+```python
+name: str = Transform.IN(env="NAME", default="Blocky")
+```
+> An optional input interface with a default value, and which additionally
+exposes the value in an environement variable `$NAME`.
+
+```python
+pypath: list[Path] = Transform.IN(env="PYTHONPATH", env_policy="append", default_factory=list)
+```
+> An optional input interface with a default which accepts a list of paths
+and exposes them in an environment variable `$PYTHONPATH`. The specified
+`env_policy` indicates the new items should be added to the env of the env
+variable if it is already set, see the section on `IEnv` for detail.
+
+```python
+result: Path = Transform.OUT()
+```
+> An automatic output interface. The output path will be automatically
+generated based on the transform directory and the interface name. This field
+cannot be set when instancing the transform.
+
+```python
+result: Path = Transform.OUT(init=True)
+```
+> A required output interface. The output path must be specified when the
+transform is instanced.
+
+```python
+result: Path = Transform.OUT(init=True, default=...)
+```
+> An optional output interface with an automatic default if a value is not
+specified. The magic value '...' can also be used with input interfaces
```

Comment:
Again kind of limited by the base dataclass field syntax. You have to set a default otherwise the field will be required, but it is really useful to be able to say "If I provide a value use that value, otherwise use the generated field".

I thought about using None, but None could be a valid value.
Could use an `auto` magic value that you import, but want to make it easy to write these without knowing about lots of special import values. 
I think the ellipsis `...` is quite nice because it's a value that's always available, can't be valid (not jsonable), and whose meaning in writing actually matches quite nicely the meaning here

> the omission from speech or writing of a word or words that are superfluous **or able to be understood from contextual clues**